### ISSUE TITLE
Alignment of eval mask for H1 dataset

### ISSUE THREAD

**Issue by trungle93** [snel-repo/falcon-challenge#26]
Hi FALCON organizers,

I'm looking at the H1 dataset and trying to understand the role of the eval_mask. My understanding was that eval_mask indicates the inter-trial periods; however, when I plotted the eval_mask together with the covariates, it doesn't seem to align with the start/end of a trial. It also doesn't seem to indicate the still times, as the covariates can be both high and low when eval_mask is 1.

Below is a minimal example to reproduce the observations. We can observe the misalignment across held-in calibration, held-in minival, and held-out calibration sets. Could you please let me know if I'm missing something here? Thanks!

```python
from pathlib import Path 
from falcon_challenge.dataloaders import load_nwb
from falcon_challenge.config import FalconTask
import matplotlib.pyplot as plt

data_dir = Path("./data/000954/")
# heldin_calib_files = sorted([f for f in data_dir.rglob('*held-in-calib*.nwb')])
# heldin_minival_files = sorted([f for f in data_dir.rglob('*held-in-minival*.nwb')])
heldout_calib_files = sorted([f for f in data_dir.rglob('*held-out-calib*.nwb')])

task = FalconTask.h1
for idx, file in enumerate(heldout_calib_files):
    neural, covariates, trial_change, eval_mask = load_nwb(file, task)

    fig, axes = plt.subplots(4, 1, figsize=(12, 10), sharex=True)

    # Plot neural data
    axes[0].imshow(neural.T, aspect='auto', cmap='viridis')
    axes[0].set_title('Neural')
    axes[0].set_ylabel('Channels')

    # Plot covariates
    for i in range(covariates.shape[1]):
        axes[1].plot(covariates[:, i], label=f'Covariate {i+1}')
    axes[1].set_title('Covariates')
    axes[1].set_ylabel('Value')
    axes[1].legend(loc='lower center', bbox_to_anchor=(0.5, -0.3), ncol=covariates.shape[1])

    # Plot trial change
    axes[2].plot(trial_change, color='magenta')
    axes[2].set_title('Trial Change')
    axes[2].set_ylabel('Change')
    axes[2].set_ylim(-0.1, 1.1)

    # Plot eval mask
    axes[3].plot(eval_mask, color='orange')
    axes[3].set_title('Evaluation Mask')
    axes[3].set_ylabel('Mask')
    axes[3].set_ylim(-0.1, 1.1)

    axes[3].set_xlabel('Time (bins)')
    plt.tight_layout()
    plt.show()
```

![Image](https://github.com/user-attachments/assets/2179126e-127e-4c1f-8a67-e979ca9f9c88)

**Comment by joel99** [comment]
Hi Trung, thanks for writing. I just took a look at the original state sequence ramp up, your example is correct and here's an alternative view. with some more privileged state information. 

1. As for why the eval. mask doesn't seem informative: 

These calibration trials follow a number of different state phases, mostly alternating between presentation and movement.
```
['Intertrial' 'Presentation' 'Reach' 'Presentation2' 'Orient' 'SnapTo'
 'Presentation3' 'Shape' 'Presentation4' 'Grasp' 'Presentation5' 'Carry'
 'Orient2' 'Presentation7' 'Release' 'Presentation6']
```

The eval mask is set be zero for the "Intertrial" and "SnapTo" phases. From the topmost panel in my plot, "Intertrial" might be a misleading name, it seems like the trials go consecutively straight from the end of one movement to the start of another, without intermediate breaks (this is not too surprising). The eval mask is really only briefly inactive for the SnapTo phase, then, in the bulk of the data. 

2. As for why the covariates seem to be active through the trial change:

Most trials should end with a delta in one of the grasp dimensions (dim 5 6 7). The covariate targets provided are given as velocities. You can see pretty good alignment of trial change with the offset of a grasp dimension in my plot. I think the smoothing done on these positions when they are turned into a velocity makes it so there's a bit of these grasp velocities occuring during the trial change. From the participant view, there's no real timing significance of the trial changing from one to the next, so I think this clipping is fine. It probably does mean the decoder should not consider the trial change signal seriously.

![Image](https://github.com/user-attachments/assets/18c09e27-7deb-4d12-8a4f-69c80835997e)

Thanks for raising these. Let me know if you'd l ike further clarification.

**Comment by trungle93** [comment]
Thank you Joel for the clarification - that was helpful! I have a couple more questions:

1. What does the subject do during the SnapTo periods and why wouldn't we want to evaluate model predictions during these times? The position and velocity look like they are active during these periods.

2. In the demos (h1.ipynb and sklearn_decoder.py), the Ridge regressor was trained and evaluated with the still times removed. My understanding is it was due to concerns that the subject is not constrained to follow the robotic arm movement in these periods and therefore the neural activity-behavior covariates relationship as we recorded does not hold anymore. It was noted in h1.ipynb that "H1, different than the monkey kinematic tasks will ONLY evaluate on active phases of the trial where there is motion, so please note the evaluation mask provided in the data". The eval_mask returned by falcon_challenge.dataloaders.load_nwb is not excluding the still times - would the models be evaluated on the still times when uploaded to EvalAI?

**Comment by joel99** [comment]
1. The SnapTo phase has the arm on screen snap from "near" the goal to "on the goal." It happens really quickly, in a way that's implausible for the subject to follow,  but the active position/velocity just reflects that the arm on screen is moving.
2. Hm. That's fair. That text and decoder may have been written at times when the eval mask was computed differently. The evaluation will be conducted on the still times, it's computed exactly based on eval mask. I'd advise just obeying eval mask.

**Comment by trungle93** [comment]
That makes sense. Thank you for helping with my questions - I'm closing the issue now.
